### PR TITLE
Fuzz testing

### PR DESCRIPTION
**Pull request by jomuel**
This PR contains a first messaging fuzz test (as requested in issue #1155) as a proposal what fuzz tests/property tests for raiden could look like.

Before this test is merged, it will be added to travis in a new job, extended to test more situations and the code will be cleaned up.

### CODE REVIEW

File: raiden/tests/unit/fuzz/test_state_changes.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+            initiator=self.address,
+            target=self.channel.partner_state.address,
+            secret=secret,
+        )
+
+    def _secret_in_use(self, secret):
+        return sha3(secret) in self.chain_state.payment_mapping.secrethashes_to_task
+
+    def _available_amount(self):
+        return channel.get_distributable(self.channel.our_state, self.channel.partner_state)
```
Comment (konradkonrad) [inline raiden/tests/unit/fuzz/test_state_changes.py]: I believe, `channel.get_distributable...` is missing a reference to `self` :)

Comment (jomuel) [reply raiden/tests/unit/fuzz/test_state_changes.py]: I don't think so. `get_distributable` is a function in the `raiden.transfer.channel` module, not a method of `self.channel`

Comment (LefterisJP) [reply raiden/tests/unit/fuzz/test_state_changes.py]: > Maybe consider direct import of the raiden.transfer.channel methods in use, to help the 

This is how we use the `channel` functions of the state machine all over the codebase.

File: raiden/tests/unit/fuzz/test_state_changes.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    def channel_state_invariants(self):
+        for netting_channel in self.channels:
+            our_state = netting_channel.our_state
+            partner_state = netting_channel.partner_state
+
+            our_balance = channel.get_balance(our_state, partner_state)
+            partner_balance = channel.get_balance(partner_state, our_state)
+
+            assert channel.get_amount_locked(our_state) <= our_balance
+            assert channel.get_amount_locked(partner_state) <= partner_balance
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden/tests/unit/fuzz/test_state_changes.py]: @hackaugusto , the constraints that we must enforce are listed here: https://raiden-network-specification.readthedocs.io/en/latest/smart_contracts.html#protocol-values-constraints. For both the smart contracts and Raiden client. Is this what you asked for?

[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/tests/unit/fuzz/test_state_changes.py]: Yes, thank you :)

File: raiden/tests/unit/fuzz/test_state_changes.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+                partner_transferred_amount = partner_state.balance_proof.transferred_amount
+                assert partner_transferred_amount >= 0
+
+            assert channel.get_distributable(our_state, partner_state) >= 0
+            assert channel.get_distributable(partner_state, our_state) >= 0
+
+            our_deposit = netting_channel.our_total_deposit
+            our_amount_locked = channel.get_amount_locked(our_state)
+            our_balance = channel.get_balance(our_state, partner_state)
+            assert 0 <= our_amount_locked <= our_balance <= our_deposit
```
Comment (hackaugusto) [inline raiden/tests/unit/fuzz/test_state_changes.py]: This is not correct:

- `our_amount_locked` may be larger then `our_deposit`, this happens when we receive a transfer partner, which gives us a balance higher then our deposit.
- `our_balance` can too be higher then our deposit, for the same reason

Comment (loredanacirstea) [reply raiden/tests/unit/fuzz/test_state_changes.py]: Yes, and as mentioned in https://raiden-network-specification.readthedocs.io/en/latest/smart_contracts.html#protocol-values-constraints, we have the following invariant: `(5.1 R) L1 <= AB1, L1 <= TAD, L1 >= 0`. Therefore, the above assertion can be split into:

```
assert 0 <= our_amount_locked <= our_balance
assert our_amount_locked <= our_deposit + partner_deposit  
```
Note that `TAD` is actually the total available deposit - if we had the withdraw implemented, we would instead have: `assert our_amount_locked <= our_deposit + partner_deposit - out_withdrawn - partner_withdrawn`. This MUST be changed at that point, so maybe worth documenting this in the [withdraw issue](https://github.com/raiden-network/raiden/issues/1498)?

Comment (jomuel) [reply raiden/tests/unit/fuzz/test_state_changes.py]: Thanks, I changed the checks accordingly.

File: raiden/tests/unit/fuzz/test_state_changes.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+            partner_deposit = netting_channel.partner_total_deposit
+            total_deposit = our_deposit + partner_deposit
+
+            our_amount_locked = channel.get_amount_locked(our_state)
+            our_balance = channel.get_balance(our_state, partner_state)
+            partner_amount_locked = channel.get_amount_locked(partner_state)
+            partner_balance = channel.get_balance(partner_state, our_state)
+
+            assert 0 <= our_amount_locked <= our_balance
+            assert 0 <= partner_amount_locked <= partner_balance
```
[the remark this answers is not in the document]
Comment (jomuel) [reply raiden/tests/unit/fuzz/test_state_changes.py]: The specs use `AB1` (available balance) as a synonym for capacity. If I'm not mistaken, that means `our_total_deposit` plus the partner's `transferred_amount` minus our own `transferred_amount`, which would match the return value of `channel.get_balance` function used here. So yes, the definitions are the same.

[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/tests/unit/fuzz/test_state_changes.py]: @jomuel I also think it's correct. `AB` seems to refer to the available balance as defined in the code [here](https://github.com/raiden-network/raiden/blob/26ffa1a1ad1d3236690daad307b573db35f7cc5a/raiden/transfer/channel.py#L852). We also got the [distributable](https://github.com/raiden-network/raiden/blob/26ffa1a1ad1d3236690daad307b573db35f7cc5a/raiden/transfer/channel.py#L889) which is the above `balance - locked_amount` but I don't think this is what the constraint is checking here.

**Comment by LefterisJP** [comment]
I have a question though. I thought that the point of fuzz testing was to create state changes with random but valid values and feed them to the state machine and see that it does not break. I am not sure how we are achieving this with this approach. Perhaps I misunderstand how hypothesis works but the `test_regression_malicious_secret_request_handled_properly` seems to test one specific scenario we know of. What am I missing?

**Comment by jomuel** [comment]
@LefterisJP Yes, `test_regression_malicious_secret_request_handled_properly` really just replays a single path that used to fail before Paul fixed issue #3000. It's code is mostly generated by Hypothesis.

Notice though that pytest will collect another test from this file, `TestInitiator` (pyunit-style test class). That is where the real fuzz test happens, it generates 100 randomized paths with randomized input data from the defined `@rules`.

**Pull request merged**